### ISSUE TITLE
[ossia-pd] some problems with (and improvements for) ossia.attribute

### ISSUE THREAD

**Issue by bltzr** [OSSIA/libossia#238]
- [ ] ossia.attributeS (plural) would be a better name, as we address several attributes with one object

- [ ] changing the clip attribute only happens one parameter by one when using pattern-matching 
(i.e. for foo.<1..4>, first time we send [clip both(, it will change it for foo.1, second time for foo.2, and so on...)

- [ ] @jcelerier proposed that all attributes (just like mute, rate and unit) be local to the remote - I think that's potentially very interesting for unit, clip and range, especially because ranges have different values for different units (BTW, once this is done, we should automatically convert the range values to the remote's unit...)
For other attributes (such as tags, description or hidden) I don't see the interest of that, but I don't see a reason for making an exception...

- [ ] also (and once this is done, too), the remote's attributes should be initialized with the values of the parameter's attributes

**Comment by avilleret** [comment]
- i'm not found of plural versi on of object name, because in that case, we should rename also other objects since they can deal with several parameters with pattern matching

- I won't do something like you describe, I prefer to add a second inlet and use it to specify which parameter to target, like for remote and parameter object

- I agree even if for me the range example doesn't make sense : converting range to right unit is something different than having local range filter

- of course, since I juste removed them, I'll put them back :-)

**Comment by bltzr** [comment]
- well, parameter defines **one** parameter that can get multiplicated with pattern-matching, here we manipulate **several** attributes that can address multiple parameters with pattern-matching, so that's a different issue

- I'm all for adding a second inlet (if that's not too time-consuming) but I'm wondering how we can reverting to addressing all instances (also for param and remote ?) - something like [all( in the 2d inlet ? - also, if we do so, then definitely want to have a dedicated outlet for addresses

- yes, converting range to right unit is something different than having local range filter, sorry for mixing two problems - let's phrase it again: it's good to be able to have a different range for the remote AND I was supposing that could help achieving having converted range for other units, but maybe I was wrong...

> of course, since I've just removed them, I'll put them back :-)
no no no, that's not what I meant - let's discuss that directly on gitter...

**Comment by bltzr** [comment]
> changing the clip attribute only happens one parameter by one when using pattern-matching
(i.e. for foo.<1..4>, first time we send [clip both(, it will change it for foo.1, second time for foo.2, and so on...)

That's not a desired behavior, but a buggy behavior that happens on the current implementation, see:
![craquage](https://user-images.githubusercontent.com/113982/30591772-7339b500-9d44-11e7-8079-90701cd69af7.gif)

**Comment by bltzr** [comment]
- [ ] add the [address( message as discussed today on gitter to address only one of the bound parameters (also for remotes and parameters)

- [ ] changing the type only does it for the first element
e.g. sending type int to ossia.attribute fo*/<tut.a*> (param \@type float) will do
attribute: address foo.1/tut.ac
attribute: type int
attribute: address foo.1/tut.ab
attribute: type float
attribute: address foo.1/tut.adabra
attribute: type float

- [ ] default, mode, enable, and repetitions do the same thing than clip (instances are addressed one after the other)
this also happens when changed from the param

- [ ] mute gives the error message: "no attribute muted" (no idea what that means)

- [ ] IMHO values should be output also when not changed (now if e.g. clip is set as free, sending clip free will output nothing - this is quite confusing!)

- [ ] when changing the range from the param, it does not change in ø.attr (even when doing getrange after) - same for min and max

All of this should be easily testable from test-ø.attr-pattern.pd in the tests folder

**Comment by avilleret** [comment]
> IMHO values should be output also when not changed (now if e.g. clip is set as free, sending clip free will output nothing - this is quite confusing!)

this is the behavior defined in libossia itself, when you set an attribute with its current value, it is not spread over the network. There is kind of repetition filter on attribute.
This behavior is OK for me even if it could be confusing the first time

**Comment by avilleret** [comment]
@bltzr could we close this ?

**Comment by bltzr** [comment]
> 
> this is the behavior defined in libossia itself, when you set an attribute with its current value, it is not spread over the network.

But here is not spread on the network, that happens in the same application.

**Comment by avilleret** [comment]
sorry, network was the wrong word.

**Comment by avilleret** [comment]
what do @jcelerier think about not filtering the attribute changes ?
